src/kms.rs

```diff
@@ -0,0 +1,174 @@
+//! Types for integrating with a Key Management Server to use with Parquet Modular Encryption
+
+use parquet::errors::Result;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::{Arc, RwLock, RwLockReadGuard};
+
+/// API for interacting with a KMS.
+/// This should be implemented by user code for integration with your KMS.
+pub trait KmsClient: Send + Sync {
+    /// Wrap encryption key bytes using the KMS with the specified master key
+    fn wrap_key(&self, key_bytes: &[u8], master_key_identifier: &str) -> Result<String>;
+
+    /// Unwrap a wrapped encryption key using the KMS with the specified master key
+    fn unwrap_key(&self, wrapped_key: &str, master_key_identifier: &str) -> Result<Vec<u8>>;
+}
+
+/// A reference-counted reference to a generic [`KmsClient`]
+pub type KmsClientRef = Arc<dyn KmsClient>;
+
+/// Holds configuration options required to connect to a KMS
+#[derive(Clone, Debug)]
+pub struct KmsConnectionConfig {
+    kms_instance_url: String,
+    kms_instance_id: String,
+    key_access_token: Arc<RwLock<String>>,
+    custom_kms_conf: HashMap<String, String>,
+}
+
+impl KmsConnectionConfig {
+    /// Create a builder for a [`KmsConnectionConfig`]
+    pub fn builder() -> KmsConnectionConfigBuilder {
+        KmsConnectionConfigBuilder::default()
+    }
+
+    /// ID of the KMS instance to use for encryption,
+    /// if multiple instances are available.
+    pub fn kms_instance_id(&self) -> &str {
+        &self.kms_instance_id
+    }
+
+    /// URL of the KMS instance to use for encryption.
+    pub fn kms_instance_url(&self) -> &str {
+        &self.kms_instance_url
+    }
+
+    /// The access token to pass to the KMS.
+    pub fn key_access_token(&self) -> String {
+        self.key_access_token.read().unwrap().clone()
+    }
+
+    /// Return the key access token inside a read lock.
+    pub(crate) fn read_key_access_token(&self) -> RwLockReadGuard<String> {
+        self.key_access_token.read().unwrap()
+    }
+
+    /// Any KMS specific configuration options
+    pub fn custom_kms_conf(&self) -> &HashMap<String, String> {
+        &self.custom_kms_conf
+    }
+
+    /// Update the authorization token to be passed to the KMS.
+    pub fn refresh_key_access_token(&self, key_access_token: String) {
```

Comment:
`set_key_access_token` would be more idiomatic and consistent with `KmsConnectionConfigBuilder.set_key_access_token`.

Reply:
I think I prefer the `refresh` name as it makes it clear this is different to a normal setter. This is intended to be used in cases where a KMS access token can expire and needs to be refreshed without having to recreate a `CryptoFactory` etc. This will cause any cached KMS clients to be invalidated and new clients will be created as needed. This isn't very clear from the doc comment though, I'll follow up and update that.

Reply:
I see, `refresh` is a better name then.